contracts/governance/Staking/Staking.sol

```diff
@@ -22,29 +22,35 @@ contract Staking is WeightedStaking{
         //stake for the msg.sender if not specified otherwise
         if(stakeFor == address(0))
             stakeFor = msg.sender;
-        require(currentBalance(stakeFor, until) == 0, "Staking:stake: use 'increaseStake' to increase an existing staked position");
-        
-        //do not stake longer than the max duration
-        if (until > block.timestamp + MAX_DURATION)
-            until = block.timestamp + MAX_DURATION;
-            
-        //retrieve the SOV tokens
-        bool success = SOVToken.transferFrom(msg.sender, address(this), amount);
-        require(success);
-        
-        //lock the tokens and update the balance by updating the user checkpoint
-        _increaseUserStake(stakeFor, until, amount);
-        
-        //increase staked token count until the new locking date
-        _increaseDailyStake(until, amount);
-        
-        //delegate to self in case no address provided
+        //delegate for stakeFor  if not specified otherwise
         if(delegatee == address(0))
-            _delegate(stakeFor, stakeFor, until);
-        else
+            delegatee = stakeFor;
+
+        if (currentBalance(stakeFor, until) > 0) {
+            //change delegate if new one is specified
             _delegate(stakeFor, delegatee, until);
-        
-        emit TokensStaked(stakeFor, amount, until, amount);
+            //increase stake, new delegate will be used
+            _increaseStake(amount, stakeFor, until);
+        } else {
+            //do not stake longer than the max duration
+            if (until > block.timestamp + MAX_DURATION)
+                until = block.timestamp + MAX_DURATION;
```

Comment:
this needs to be done before the if-else. It could be that the user already has tokens staked until that point. 

additional optional improvement suggestion:
The same code which is in the else is also part of _increaseStake. I think we can avoid code duplication by calling ```_icreaseStake``` in any event and just let it contain the steps whcih are the same for both functions: ```increaseStake``` and ```stake```.

Reply:
oh, thanks

refactored